File: day2/typescript/day2.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

const maximum = R.apply(Math.max);
const minimum = R.apply(Math.min);
const mapSum = (fn: (x: number[]) => number) => R.compose(R.sum, R.map(fn));

const part1 = mapSum(
    R.lift(R.subtract)(maximum, minimum),
);

const part2 = mapSum(row =>
    R.compose<number[], number[], number[], number>(
        maximum,
        R.filter<number>(x => x % 1 === 0),
        R.lift(R.divide)(row),
    )(row),
);

const readGrid = R.compose(
    R.map(R.map(parseInt)),
    R.map(R.split('\t')),
    R.split('\n'),
);

async function day2() {
    const input = await promisify(readFile)('day2/input.txt', 'utf8');
    const grid = readGrid(input);

    console.log(part1(grid));
    console.log(part2(grid));
}

day2();

File: day2/haskell/day2.hs
import Control.Arrow ((&&&))
import Data.List (subsequences)
import Data.List.Split (splitOn)
import Data.Tuple (swap)

readGrid :: String -> IO [[Int]]
readGrid s = map (map read . splitOn "\t") . lines <$> readFile s

seqPairs :: [a] -> [(a, a)]
seqPairs = map fn . filter ((== 2) . length) . subsequences
    where fn [a, b] = (a, b)

allPairs :: [a] -> [(a, a)]
allPairs xs = ps ++ map swap ps
    where ps = seqPairs xs

mapSum :: Num b => (a -> b) -> [a] -> b
mapSum f = sum . map f

part1 :: [Int] -> Int
part1 = uncurry subtract . (minimum &&& maximum)

part2 :: [Int] -> Int
part2 = fst . head . filter ((== 0) . snd) . map (uncurry divMod) . allPairs

main :: IO ()
main = do
    grid <- readGrid "day2/input.txt"
    print . mapSum part1 $ grid
    print . mapSum part2 $ grid